pkg/discovery/network/generic.go

```
@@ -62,13 +50,112 @@ func discoverGenericNetwork(clientSet kubernetes.Interface) (*ClusterNetwork, er
+	// expected msg is below:
+	//   "The Service \"invalid-svc\" is invalid: spec.clusterIPs: Invalid value: []string{\"1.1.1.1\"}:
+	//   failed to allocated ip:1.1.1.1 with error:provided IP is not in the valid range.
+	//   The range of valid IPs is 10.45.0.0/16"
+	// expected matched string is below:
+	//   10.45.0.0/16
+	re := regexp.MustCompile(".*valid IPs is (.*)$")
+
+	match := re.FindStringSubmatch(msg)
+	if match == nil {
+		return "", fmt.Errorf("parsing (%s) failed. it doesn't match with %v", msg, re)
+	}
+
+	// returns first matching string
+	return match[1], nil
+}
+
+func findPodIPRange(clientSet kubernetes.Interface) (string, error) {
+	podIPRange, err := findPodIPRangeKubeController(clientSet)
+	if err == nil && podIPRange != "" {
+		return podIPRange, nil
+	}
+
+	podIPRange, err = findPodIPRangeKubeProxy(clientSet)
+	if err == nil && podIPRange != "" {
+		return podIPRange, nil
+	}
+
+	podIPRange, err = findPodIPRangeNodeSpec(clientSet)
+	if err == nil && podIPRange != "" {
+		return podIPRange, nil
+	}
+
+	return "", fmt.Errorf("podIPRange can't be found by any methods")
+}
+
 func findPodIPRangeKubeController(clientSet kubernetes.Interface) (string, error) {
 	return findPodCommandParameter(clientSet, "component=kube-controller-manager", "--cluster-cidr")
 }
 
 func findPodIPRangeKubeProxy(clientSet kubernetes.Interface) (string, error) {
 	return findPodCommandParameter(clientSet, "component=kube-proxy", "--cluster-cidr")
 }
+
+func findPodIPRangeNodeSpec(clientSet kubernetes.Interface) (string, error) {
+	nodes, err := clientSet.CoreV1().Nodes().List(v1meta.ListOptions{})
+
+	if err != nil {
+		return "", errors.WithMessagef(err, "error listing nodes")
+	}
+
+	return parseToPodCidr(nodes.Items)
+}
+
+func parseToPodCidr(nodes []v1.Node) (string, error) {
+	for _, node := range nodes {
+		if node.Spec.PodCIDR != "" {
+			return node.Spec.PodCIDR, nil
+		}
+	}
+
+	return "", fmt.Errorf("no node with Spec.PodCIDR found")
```

I don't think we should return an error here if not found. We should let the caller handle not found as it sees fit (it so happens this is the last resort for the caller but that could change).